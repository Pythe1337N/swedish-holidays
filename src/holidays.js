Date.prototype.plusDays = function (d) {
  const date = new Date(this);
  date.setDate(this.getDate() + (d || 0));
  return date;
};

Date.prototype.plusWeeks = function (d) {
  const date = new Date(this);
  date.setDate(this.getDate() + (d || 0) * 7);
  return date;
};

Date.prototype.plusYears = function (d) {
  const date = new Date(this);
  date.setFullYear(this.getFullYear() + (d || 0));
  return date;
};

const swedishHolidayNames = {
  newYearsDay: "Nyårsdagen",
  twelfthNight: "Trettondagsafton",
  epiphany: "Trettondedag jul",
  maundyThursday: "Skärtorsdagen",
  goodFriday: "Långfredagen",
  holySaturday: "Påskafton",
  easterSunday: "Påskdagen",
  easterMonday: "Annandag påsk",
  walpurgisNight: "Valborgsmässoafton",
  mayFirst: "Första maj",
  ascensionDay: "Kristi himmelsfärdsdag",
  swedishNationalDay: "Sveriges nationaldag",
  pentecostEve: "Pingstafton",
  whitSunday: "Pingstdagen",
  midsummerEve: "Midsommarafton",
  midsummerDay: "Midsommardagen",
  allSaintsEve: "Allhelgonaafton",
  allSaintsDay: "Alla helgons dag",
  christmasEve: "Julafton",
  christmasDay: "Juldagen",
  boxingDay: "Annandag jul",
  newYearsEve: "Nyårsafton",
};

const FRIDAY = 5;

const isHoliday = (date, language) => {
  if (!date) {
    return isHoliday(new Date());
  }
  const holidays = getHolidaysForYear(date.getFullYear(), language);
  const [holiday] = holidays.filter(
    (holiday) =>
      holiday.day === date.getDate() &&
      holiday.month === date.getMonth() + 1 &&
      holiday.year === date.getFullYear()
  );
  return holiday || false;
};

const getHolidayJSON = (name, date, isPublicHoliday = false) => ({
  name,
  date,
  day: date.getDate(),
  month: date.getMonth() + 1,
  year: date.getFullYear(),
  isPublicHoliday,
});

const fixedDate = (year, month, day) => {
  const m =
    month.toString().length > 1 ? month.toString() : "0" + month.toString();
  const d = day.toString().length > 1 ? day.toString() : "0" + day.toString();
  return new Date(year + "-" + m + "-" + d + "T00:00:00Z");
};

const firstOfWeekdayAfterDate = (weekday, refDate) => {
  return refDate.plusDays(
    (refDate.getDay() > weekday ? 7 : 0) - refDate.getDay() + weekday
  );
};

const getUpcomingHolidays = (language) => {
  const now = new Date();
  const nextYear = new Date().plusYears(1);
  return getHolidaysForYear(now.getFullYear(), language)
    .concat(getHolidaysForYear(nextYear.getFullYear(), language))
    .filter(
      (holiday) =>
        holiday.date.getTime() >= now.getTime() &&
        holiday.date.getTime() < nextYear.getTime()
    );
};

const getHolidaysForYear = (year, language) => {
  if (!year) {
    return getHolidaysForYear(new Date().getFullYear(), language);
  }
  if (!language) {
    return getHolidaysForYear(year, swedishHolidayNames);
  }
  const holidays = [];
  if (year >= 1582 && year <= 8702) {
    const goldenNumber = Math.floor((year % 19) + 1);
    const century = Math.floor(year / 100 + 1);
    const corx = Math.floor((3 * century) / 4 - 12);
    const corz = Math.floor((8 * century + 5) / 25 - 5);
    const sunday = Math.floor((5 * year) / 4 - corx - 10);
    let epact = Math.floor((11 * goldenNumber + 20 + corz - corx + 30) % 30);

    if ((epact === 25 && goldenNumber > 11) || epact === 24) {
      epact += 1;
    }

    let moon = 44 - epact;

    if (moon < 21) {
      moon += 30;
    }

    moon = moon + 7 - ((sunday + moon) % 7);

    let month = 3;
    let day = moon;

    if (moon > 31) {
      month = 4;
      day = moon - 31;
    }

    const easterSunday = fixedDate(year, month, day);
    const christSkyFly = easterSunday.plusWeeks(6).plusDays(-3);
    const pingst = easterSunday.plusWeeks(7);

    holidays.push(
      getHolidayJSON(language.maundyThursday, easterSunday.plusDays(-3))
    );
    holidays.push(
      getHolidayJSON(language.goodFriday, easterSunday.plusDays(-2), true)
    );
    holidays.push(
      getHolidayJSON(language.holySaturday, easterSunday.plusDays(-1))
    );
    holidays.push(getHolidayJSON(language.easterSunday, easterSunday, true));
    holidays.push(
      getHolidayJSON(language.easterMonday, easterSunday.plusDays(1), true)
    );

    holidays.push(getHolidayJSON(language.ascensionDay, christSkyFly, true));

    holidays.push(getHolidayJSON(language.pentecostEve, pingst.plusDays(-1)));
    holidays.push(getHolidayJSON(language.whitSunday, pingst, true));
  }

  const fylla = firstOfWeekdayAfterDate(FRIDAY, fixedDate(year, 6, 19));
  const bakis = fylla.plusDays(1);

  holidays.push(getHolidayJSON(language.midsummerEve, fylla));
  holidays.push(getHolidayJSON(language.midsummerDay, bakis, true));

  const spooky = firstOfWeekdayAfterDate(FRIDAY, fixedDate(year, 10, 30));
  holidays.push(getHolidayJSON(language.allSaintsEve, spooky));
  holidays.push(
    getHolidayJSON(language.allSaintsDay, spooky.plusDays(1), true)
  );

  holidays.push(
    getHolidayJSON(language.newYearsDay, fixedDate(year, 1, 1), true)
  );
  holidays.push(getHolidayJSON(language.twelfthNight, fixedDate(year, 1, 5)));
  holidays.push(getHolidayJSON(language.epiphany, fixedDate(year, 1, 6), true));
  holidays.push(
    getHolidayJSON(language.walpurgisNight, fixedDate(year, 4, 30))
  );
  holidays.push(getHolidayJSON(language.mayFirst, fixedDate(year, 5, 1), true));
  holidays.push(
    getHolidayJSON(language.swedishNationalDay, fixedDate(year, 6, 6), true)
  );
  holidays.push(getHolidayJSON(language.christmasEve, fixedDate(year, 12, 24)));
  holidays.push(
    getHolidayJSON(language.christmasDay, fixedDate(year, 12, 25), true)
  );
  holidays.push(
    getHolidayJSON(language.boxingDay, fixedDate(year, 12, 26), true)
  );
  holidays.push(getHolidayJSON(language.newYearsEve, fixedDate(year, 12, 31)));
  return holidays.sort((a, b) => a.date.getTime() - b.date.getTime());
};

module.exports = {
  getHolidays: getHolidaysForYear,
  getUpcomingHolidays,
  isHoliday,
  language: { ...swedishHolidayNames },
};
