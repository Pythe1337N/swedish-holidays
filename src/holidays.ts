import { swedishHolidayNames, sunday, LanguageType } from "./lang";
import {
  addDays,
  addWeeks,
  addYears,
  firstOfWeekdayAfterDate,
  fixedDate,
  getHolidayJSON,
  HolidayType,
} from "./lib";

const FRIDAY = 5;
const SUNDAY = 0;

export const language = swedishHolidayNames;

export const getHoliday = (
  date: Date,
  language?: LanguageType
): HolidayType | undefined => {
  if (!date) {
    return getHoliday(new Date());
  }
  const holidays = getHolidays(date.getFullYear(), language);
  return holidays.find(
    (holiday) =>
      holiday.day === date.getDate() &&
      holiday.month === date.getMonth() + 1 &&
      holiday.year === date.getFullYear()
  );
};

export const getPublicHoliday = (
  date: Date,
  language?: LanguageType
): HolidayType | undefined => {
  if (!date) {
    return getPublicHoliday(new Date());
  }

  if (date.getDay() === SUNDAY) {
    return getHolidayJSON(sunday, date, true);
  }

  const holidays = getHolidays(date.getFullYear(), language);
  return holidays.find(
    (holiday) =>
      holiday.day === date.getDate() &&
      holiday.month === date.getMonth() + 1 &&
      holiday.year === date.getFullYear() &&
      holiday.isPublicHoliday
  );
};

export const isHoliday = (date: Date, language?: LanguageType): boolean =>
  Boolean(getHoliday(date, language));

export const isPublicHoliday = (date: Date, language?: LanguageType): boolean =>
  Boolean(getPublicHoliday(date, language));

export const getUpcomingHolidays = (language: LanguageType) => {
  const now = new Date();
  const nextYear = addYears(now, 1);
  return getHolidays(now.getFullYear(), language)
    .concat(getHolidays(nextYear.getFullYear(), language))
    .filter(
      (holiday) =>
        holiday.date.getTime() >= now.getTime() &&
        holiday.date.getTime() < nextYear.getTime()
    );
};

export const getHolidays = (
  year: number,
  language?: Record<string, string>
): HolidayType[] => {
  if (!year) {
    return getHolidays(new Date().getFullYear(), language);
  }
  if (!language) {
    return getHolidays(year, swedishHolidayNames);
  }
  const holidays: HolidayType[] = [];
  if (year >= 1582 && year <= 8702) {
    const goldenNumber = Math.floor((year % 19) + 1);
    const century = Math.floor(year / 100 + 1);
    const corx = Math.floor((3 * century) / 4 - 12);
    const corz = Math.floor((8 * century + 5) / 25 - 5);
    const sunday = Math.floor((5 * year) / 4 - corx - 10);
    let epact = Math.floor((11 * goldenNumber + 20 + corz - corx + 30) % 30);

    if ((epact === 25 && goldenNumber > 11) || epact === 24) {
      epact += 1;
    }

    let moon = 44 - epact;

    if (moon < 21) {
      moon += 30;
    }

    moon = moon + 7 - ((sunday + moon) % 7);

    let month = 3;
    let day = moon;

    if (moon > 31) {
      month = 4;
      day = moon - 31;
    }

    const easterSunday = fixedDate(year, month, day);
    const christSkyFly = addDays(addWeeks(easterSunday, 6), -3);
    const pingst = addWeeks(easterSunday, 7);

    holidays.push(
      getHolidayJSON(language.maundyThursday, addDays(easterSunday, -3))
    );
    holidays.push(
      getHolidayJSON(language.goodFriday, addDays(easterSunday, -2), true)
    );
    holidays.push(
      getHolidayJSON(language.holySaturday, addDays(easterSunday, -1))
    );
    holidays.push(getHolidayJSON(language.easterSunday, easterSunday, true));
    holidays.push(
      getHolidayJSON(language.easterMonday, addDays(easterSunday, 1), true)
    );

    holidays.push(getHolidayJSON(language.ascensionDay, christSkyFly, true));

    holidays.push(getHolidayJSON(language.pentecostEve, addDays(pingst, -1)));
    holidays.push(getHolidayJSON(language.whitSunday, pingst, true));
  }

  const midsummerEve = firstOfWeekdayAfterDate(FRIDAY, fixedDate(year, 6, 19));
  const midsummerDay = addDays(midsummerEve, 1);

  holidays.push(getHolidayJSON(language.midsummerEve, midsummerEve));
  holidays.push(getHolidayJSON(language.midsummerDay, midsummerDay, true));

  const allSaintsEve = firstOfWeekdayAfterDate(FRIDAY, fixedDate(year, 10, 30));
  holidays.push(getHolidayJSON(language.allSaintsEve, allSaintsEve));
  holidays.push(
    getHolidayJSON(language.allSaintsDay, addDays(allSaintsEve, 1), true)
  );

  holidays.push(
    getHolidayJSON(language.newYearsDay, fixedDate(year, 1, 1), true)
  );
  holidays.push(getHolidayJSON(language.twelfthNight, fixedDate(year, 1, 5)));
  holidays.push(getHolidayJSON(language.epiphany, fixedDate(year, 1, 6), true));
  holidays.push(
    getHolidayJSON(language.walpurgisNight, fixedDate(year, 4, 30))
  );
  holidays.push(getHolidayJSON(language.mayFirst, fixedDate(year, 5, 1), true));
  holidays.push(
    getHolidayJSON(language.swedishNationalDay, fixedDate(year, 6, 6), true)
  );
  holidays.push(getHolidayJSON(language.christmasEve, fixedDate(year, 12, 24)));
  holidays.push(
    getHolidayJSON(language.christmasDay, fixedDate(year, 12, 25), true)
  );
  holidays.push(
    getHolidayJSON(language.boxingDay, fixedDate(year, 12, 26), true)
  );
  holidays.push(getHolidayJSON(language.newYearsEve, fixedDate(year, 12, 31)));
  return holidays.sort((a, b) => {
    return a.date.getTime() - b.date.getTime();
  });
};
