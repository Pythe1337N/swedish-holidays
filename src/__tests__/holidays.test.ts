import { test, expect } from "vitest";
import { isPublicHoliday, getHolidays, language } from "../holidays";

const expectations = [
  {
    name: "Nyårsdagen",
    day: 1,
    month: 1,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Trettondagsafton",
    day: 5,
    month: 1,
    year: 2020,
    isPublicHoliday: false,
  },
  {
    name: "Trettondedag jul",
    day: 6,
    month: 1,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Skärtorsdagen",
    day: 9,
    month: 4,
    year: 2020,
    isPublicHoliday: false,
  },
  {
    name: "Långfredagen",
    day: 10,
    month: 4,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Påskafton",
    day: 11,
    month: 4,
    year: 2020,
    isPublicHoliday: false,
  },
  {
    name: "Påskdagen",
    day: 12,
    month: 4,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Annandag påsk",
    day: 13,
    month: 4,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Valborgsmässoafton",
    day: 30,
    month: 4,
    year: 2020,
    isPublicHoliday: false,
  },
  {
    name: "Första maj",
    day: 1,
    month: 5,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Kristi himmelsfärdsdag",
    day: 21,
    month: 5,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Pingstafton",
    day: 30,
    month: 5,
    year: 2020,
    isPublicHoliday: false,
  },
  {
    name: "Pingstdagen",
    day: 31,
    month: 5,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Sveriges nationaldag",
    day: 6,
    month: 6,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Midsommarafton",
    day: 19,
    month: 6,
    year: 2020,
    isPublicHoliday: false,
  },
  {
    name: "Midsommardagen",
    day: 20,
    month: 6,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Allhelgonaafton",
    day: 30,
    month: 10,
    year: 2020,
    isPublicHoliday: false,
  },
  {
    name: "Alla helgons dag",
    day: 31,
    month: 10,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Julafton",
    day: 24,
    month: 12,
    year: 2020,
    isPublicHoliday: false,
  },
  {
    name: "Juldagen",
    day: 25,
    month: 12,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Annandag jul",
    day: 26,
    month: 12,
    year: 2020,
    isPublicHoliday: true,
  },
  {
    name: "Nyårsafton",
    day: 31,
    month: 12,
    year: 2020,
    isPublicHoliday: false,
  },
];

test("Return a nice list with all holidays for 2020", () => {
  const holidays = getHolidays(2020);
  expect(holidays).toBeDefined();
  expect(holidays.length).toEqual(expectations.length);

  for (let i = 0; i < holidays.length; i++) {
    expect(expectations[i].day).toEqual(holidays[i].day);
    expect(expectations[i].month).toEqual(holidays[i].month);
    expect(expectations[i].year).toEqual(holidays[i].year);
    expect(expectations[i].name).toEqual(holidays[i].name);
    expect(expectations[i].isPublicHoliday).toEqual(
      holidays[i].isPublicHoliday
    );
  }
  expect(holidays).toMatchSnapshot();
});

test("Identify Sunday as a public holiday", () => {
  const saturdayIsNotPublicHoliday = isPublicHoliday(new Date("2021-11-13"));
  const sundayIsPublicHoliday = isPublicHoliday(new Date("2021-11-14"));
  const mondayIsNotPublicHoliday = isPublicHoliday(new Date("2021-11-15"));
  expect(saturdayIsNotPublicHoliday).toBeFalsy();
  expect(sundayIsPublicHoliday).toBeTruthy();
  expect(mondayIsNotPublicHoliday).toBeFalsy();
});

test("Identify a public holiday as a public holiday", () => {
  const christmasEve = isPublicHoliday(new Date("2021-12-24"));
  const christmasDay = isPublicHoliday(new Date("2021-12-25"));

  expect(christmasEve).toBeFalsy();
  expect(christmasDay).toBeTruthy();
});

test("Handle translations", () => {
  const translation = { ...language };
  translation.christmasEve = "Christmas Eve";
  const [christmas] = getHolidays(2020, translation).filter(
    (holiday) => holiday.day === 24 && holiday.month === 12
  );
  expect(christmas).toBeDefined();
  expect(christmas.name).toEqual("Christmas Eve");
});
