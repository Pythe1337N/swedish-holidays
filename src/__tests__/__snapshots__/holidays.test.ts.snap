// Vitest Snapshot v1

exports[`Return a nice list with all holidays for 2020 1`] = `
[
  {
    "date": 2020-01-01T00:00:00.000Z,
    "day": 1,
    "isPublicHoliday": true,
    "month": 1,
    "name": "Nyårsdagen",
    "year": 2020,
  },
  {
    "date": 2020-01-05T00:00:00.000Z,
    "day": 5,
    "isPublicHoliday": false,
    "month": 1,
    "name": "Trettondagsafton",
    "year": 2020,
  },
  {
    "date": 2020-01-06T00:00:00.000Z,
    "day": 6,
    "isPublicHoliday": true,
    "month": 1,
    "name": "Trettondedag jul",
    "year": 2020,
  },
  {
    "date": 2020-04-09T00:00:00.000Z,
    "day": 9,
    "isPublicHoliday": false,
    "month": 4,
    "name": "Skärtorsdagen",
    "year": 2020,
  },
  {
    "date": 2020-04-10T00:00:00.000Z,
    "day": 10,
    "isPublicHoliday": true,
    "month": 4,
    "name": "Långfredagen",
    "year": 2020,
  },
  {
    "date": 2020-04-11T00:00:00.000Z,
    "day": 11,
    "isPublicHoliday": false,
    "month": 4,
    "name": "Påskafton",
    "year": 2020,
  },
  {
    "date": 2020-04-12T00:00:00.000Z,
    "day": 12,
    "isPublicHoliday": true,
    "month": 4,
    "name": "Påskdagen",
    "year": 2020,
  },
  {
    "date": 2020-04-13T00:00:00.000Z,
    "day": 13,
    "isPublicHoliday": true,
    "month": 4,
    "name": "Annandag påsk",
    "year": 2020,
  },
  {
    "date": 2020-04-30T00:00:00.000Z,
    "day": 30,
    "isPublicHoliday": false,
    "month": 4,
    "name": "Valborgsmässoafton",
    "year": 2020,
  },
  {
    "date": 2020-05-01T00:00:00.000Z,
    "day": 1,
    "isPublicHoliday": true,
    "month": 5,
    "name": "Första maj",
    "year": 2020,
  },
  {
    "date": 2020-05-21T00:00:00.000Z,
    "day": 21,
    "isPublicHoliday": true,
    "month": 5,
    "name": "Kristi himmelsfärdsdag",
    "year": 2020,
  },
  {
    "date": 2020-05-30T00:00:00.000Z,
    "day": 30,
    "isPublicHoliday": false,
    "month": 5,
    "name": "Pingstafton",
    "year": 2020,
  },
  {
    "date": 2020-05-31T00:00:00.000Z,
    "day": 31,
    "isPublicHoliday": true,
    "month": 5,
    "name": "Pingstdagen",
    "year": 2020,
  },
  {
    "date": 2020-06-06T00:00:00.000Z,
    "day": 6,
    "isPublicHoliday": true,
    "month": 6,
    "name": "Sveriges nationaldag",
    "year": 2020,
  },
  {
    "date": 2020-06-19T00:00:00.000Z,
    "day": 19,
    "isPublicHoliday": false,
    "month": 6,
    "name": "Midsommarafton",
    "year": 2020,
  },
  {
    "date": 2020-06-20T00:00:00.000Z,
    "day": 20,
    "isPublicHoliday": true,
    "month": 6,
    "name": "Midsommardagen",
    "year": 2020,
  },
  {
    "date": 2020-10-30T00:00:00.000Z,
    "day": 30,
    "isPublicHoliday": false,
    "month": 10,
    "name": "Allhelgonaafton",
    "year": 2020,
  },
  {
    "date": 2020-10-31T00:00:00.000Z,
    "day": 31,
    "isPublicHoliday": true,
    "month": 10,
    "name": "Alla helgons dag",
    "year": 2020,
  },
  {
    "date": 2020-12-24T00:00:00.000Z,
    "day": 24,
    "isPublicHoliday": false,
    "month": 12,
    "name": "Julafton",
    "year": 2020,
  },
  {
    "date": 2020-12-25T00:00:00.000Z,
    "day": 25,
    "isPublicHoliday": true,
    "month": 12,
    "name": "Juldagen",
    "year": 2020,
  },
  {
    "date": 2020-12-26T00:00:00.000Z,
    "day": 26,
    "isPublicHoliday": true,
    "month": 12,
    "name": "Annandag jul",
    "year": 2020,
  },
  {
    "date": 2020-12-31T00:00:00.000Z,
    "day": 31,
    "isPublicHoliday": false,
    "month": 12,
    "name": "Nyårsafton",
    "year": 2020,
  },
]
`;
